packages/navigator/src/browser/navigator-editor-sync.ts

```diff
@@ -0,0 +1,165 @@
+ * You may obtain a copy of the License at http://www.apache.org/licenses/LICENSE-2.0
+ */
+
+import { injectable, inject, postConstruct } from "inversify";
+import { Disposable } from '@theia/core/lib/common';
+import { EditorManager, EditorWidget } from "@theia/editor/lib/browser";
+import { ITreeNode, ISelectableTreeNode, IExpandableTreeNode, PreferenceChangeEvent } from "@theia/core/lib/browser";
+import { FileNavigatorModel } from "./navigator-model";
+import { NavigatorConfiguration, NavigatorPreferences } from "./navigator-preferences";
+
+@injectable()
+export class NavigatorEditorSynchronizer {
+
+    @inject(FileNavigatorModel)
+    private readonly fileNavigatorModel: FileNavigatorModel;
+
+    @inject(EditorManager)
+    private readonly editorManager: EditorManager;
+
+    @inject(NavigatorPreferences)
+    private readonly navigatorPreferences: NavigatorPreferences;
+
+    private activeEditorChangedSubscription: Disposable | undefined;
+
+    @postConstruct()
+    protected async init() {
+        await this.navigatorPreferences.ready;
+        this.linkWithEditorPreferenceChanged(this.navigatorPreferences['navigator.linkWithEditor']);
+        this.navigatorPreferences.onPreferenceChanged(preference => this.onPreferenceChangedHandler(preference));
+    }
+
+    private onPreferenceChangedHandler(preference: PreferenceChangeEvent<NavigatorConfiguration>) {
+        if (preference.preferenceName === 'navigator.linkWithEditor') {
+            this.linkWithEditorPreferenceChanged(preference.newValue);
+        }
+    }
+
+    private linkWithEditorPreferenceChanged(linkWithEditorNewValue: boolean | undefined) {
+        if (linkWithEditorNewValue) {
+            if (this.activeEditorChangedSubscription === undefined) {
+                this.activeEditorChangedSubscription = this.editorManager.onActiveEditorChanged(
+                    (editor: EditorWidget | undefined) => this.selectNodeByEditor(editor));
+            }
+        } else {
+            if (this.activeEditorChangedSubscription !== undefined) {
+                this.activeEditorChangedSubscription.dispose();
+                this.activeEditorChangedSubscription = undefined;
+            }
+        }
+    }
+
+    /**
+     * Reveals and selects corresponding to given editor node in the navigator.
+     * If editor is undefined nothing happens.
+     *
+     * @param editor editor widget object to reveal its node in the navigator
+     */
+    selectNodeByEditor(editor: EditorWidget | undefined) {
+        if (editor) {
+            this.selectNodeById(this.editorIdToNavigatorNodeId(editor.id));
+        }
+    }
+
+    /**
+     * Converts editor id to navigator node id.
+     * Example: 'code-editor-opener:file:///home/user/workspace/README.md' => 'file:///home/user/workspace/README.md'
+     *
+     * @param editorId id of editor tab
+     * @returns id of corresponding navigator node
+     */
+    private editorIdToNavigatorNodeId(editorId: string) {
```

Comment:
The id is not a contract, so relying on it is fragile.

Reply:
Yes, agree this is fragile part. I'll try to replace it